stdlib/src/memory/box.mojo

```diff
@@ -0,0 +1,128 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+from memory import UnsafePointer, stack_allocation, memcpy
+
+
+struct Box[T: AnyType]:
+    """A safe, owning, smart pointer.
+
+    This smart pointer is designed for cases where there is clear ownership
+    of the underlying data, and restricts access to it through the lifetime
+    system such that no more than one mutable alias for the underlying data
+    may exist.
+
+    Parameters:
+        T: The type to be stored in the Box[].
+    """
+
+    var _inner: UnsafePointer[T, AddressSpace.GENERIC]
+
+    fn __init__[T: Movable](inout self: Box[T], owned value: T):
+        """Construct a new Box[] by moving the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store. It is restricted to `Movable` here to allow efficient move construction.
+
+        Args:
+            value: The value to move into the Box[].
+        """
+        self._inner = UnsafePointer[T].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[T: ExplicitlyCopyable](inout self: Box[T], *, copy_value: T):
+        """Construct a new Box[] by explicitly copying the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_value: The value to explicitly copy into the Box[].
+        """
+        self._inner = UnsafePointer[T].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    # TODO: disambiguation and other niceties
```

Comment:
Totally fine to deal with this in a follow-up, I think we can disambiguate by providing a phony parameter (e.g. `U: NoneType = None`) to deal with overload fitness but still allowing this.